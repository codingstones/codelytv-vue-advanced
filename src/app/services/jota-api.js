import { gigService } from './jota-instances'

let fakeDays = gigService.retrieveNextGigs()

const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

export function retrieveDays() {
  return delay(2000).then(() => fakeDays)
}

export function retrieveAGig(gigId) {
  return delay(1000).then(() => gigService.retrieveAGig(gigId))
}

export function createGig(payload) {
  addGigToFakeSample(payload)
  if (!payload.hasOwnProperty('day')) {
    return Promise.reject(Error('SERVER ERROR'))
  }
  return delay(1000).then(() => payload)
}

function addGigToFakeSample(gig) {
  if (fakeDays[gig.day]) {
    fakeDays[gig.day].gigs.unshift(gig)
  }
  else {
    fakeDays[gig.day] = { date: gig.day, gigs: [gig] }
  }
}
