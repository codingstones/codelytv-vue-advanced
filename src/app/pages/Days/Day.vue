<template>
  <div v-if="!isLoading">
    <div class='day'>
      <q-card-title>{{dateLabel}}</q-card-title>

      <div v-for="gig in day.gigs">
        <GigRow :gig ="gig" :onClick="goTo"/>
      </div>
    </div>
  </div>
</template>

<script>
  import { localizedFromIso } from '../../services/date-utils'

  export default {
    props: {day: Object, isLoading: Boolean},
    methods: {
      goTo(gig) {
        this.jotaRouter.navigateToGig(gig.id)
      }
    },
    computed: {
      dateLabel() {
        return localizedFromIso(this.day.date)
      }
    }
  }
</script>

<style scoped>
  .q-card-title {
    color: black;
    font-weight: bold;
    font-size: 150%;
  }
</style>
