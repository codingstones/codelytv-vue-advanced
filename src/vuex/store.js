import Vue from 'vue'
import Vuex from 'vuex'
import { CREATE_GIG, RETRIEVE_DAYS } from '../app/services/jota-commands'
import { buildRetrieveDaysAction } from '../app/pages/Days/days-actions'
import {
  CREATE_GIG_ERROR,
  CREATE_GIG_REQUEST, CREATE_GIG_SUCCESS, createGigError, createGigRequest,
  createGigSuccess
} from '../app/pages/NewGig/new-gig-mutations'
import {
  FETCH_DAYS_ERROR, FETCH_DAYS_REQUEST, FETCH_DAYS_SUCCESS, fetchDaysError,
  fetchDaysRequest, fetchDaysSuccess
} from '../app/pages/Days/days-mutations'
import { buildCreateGigAction } from '../app/pages/NewGig/new-gig-actions'
import { gigsToday, gigsTomorrow, numberOfGigsToday, numberOfGigsTomorrow } from './getters'

Vue.use(Vuex)

export const buildInitialState = () => {
  return {
    days: {},
    error: false,
    loading: false
  }
}

export const initialState = {
  days: {},
  error: false,
  loading: false
}

export const actions = {
  [RETRIEVE_DAYS]: buildRetrieveDaysAction(),
  [CREATE_GIG]: buildCreateGigAction()
}

export const mutations = {
  [CREATE_GIG_REQUEST]: createGigRequest,
  [CREATE_GIG_SUCCESS]: createGigSuccess,
  [CREATE_GIG_ERROR]: createGigError,
  [FETCH_DAYS_REQUEST]: fetchDaysRequest,
  [FETCH_DAYS_SUCCESS]: fetchDaysSuccess,
  [FETCH_DAYS_ERROR]: fetchDaysError
}

export const getters = {
  numberOfGigsToday: numberOfGigsToday,
  numberOfGigsTomorrow: numberOfGigsTomorrow,
  gigsToday: gigsToday,
  gigsTomorrow: gigsTomorrow,
  paramId: state => state.route.params.id
}

export const storeDefinition = {
  state: buildInitialState(),
  actions,
  mutations,
  getters
}

export const store = new Vuex.Store(storeDefinition)
