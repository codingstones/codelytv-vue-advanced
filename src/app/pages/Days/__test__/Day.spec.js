import Day from '@/app/pages/Days/Day.vue'
import { storeDefinition } from '@/vuex/store'
import { renderComponent } from '@test/render-utils'
import userEvent from '@testing-library/user-event'
import { localizedFromIso } from '../../../services/date-utils'
import { FIRST_DAY } from '../../../services/__mocks__/gigs-sample'

it('renders gig date in spanish format', async() => {
  const {findByText} = renderComponent(Day, {
    store: storeDefinition,
    props: { day: FIRST_DAY, isLoading: false, onClick: jest.fn() }
  })

  expect(await findByText(localizedFromIso(FIRST_DAY.date))).toBeInTheDocument()
})

describe('When clicking buttons', () => {
  it('navigates to first gig detail', async () => {
    const FIRST_GIG = FIRST_DAY.gigs[0]
    const navigateToGigSpy = jest.fn()
    const screen = renderDays({navigateToGig: navigateToGigSpy})

    await screen.openGig(FIRST_GIG.title)

    expect(navigateToGigSpy).toHaveBeenCalledWith(FIRST_GIG.id)
  })

  it('navigates to second gig detail', async () => {
    const SECOND_GIG = FIRST_DAY.gigs[1]
    const navigateToGigSpy = jest.fn()
    const screen = renderDays({navigateToGig: navigateToGigSpy})

    await screen.openGig(SECOND_GIG.title)

    expect(navigateToGigSpy).toHaveBeenCalledWith(SECOND_GIG.id)
  })
})

function renderDays(jotaRouterInstance) {
  const screen = renderComponent(Day, {
    props: {day: FIRST_DAY, isLoading: false},
    jotaRouter: () => jotaRouterInstance
  })
  const openGig = async title => userEvent.click((await screen.findByText(title)))

  return {...screen, openGig}
}
