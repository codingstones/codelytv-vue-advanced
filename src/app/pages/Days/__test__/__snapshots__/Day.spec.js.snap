// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day matches full snapshot 1`] = `
<div>
    <div class="day">
        <div class="q-card-primary q-card-container row no-wrap">
            <div class="col column">
                <div class="q-card-title">Lunes, 18 De Septiembre</div>
                <div class="q-card-subtitle"></div>
            </div>
            <div class="col-auto self-center q-card-title-extra"></div>
        </div>
        <div>
            <div class="q-card">
                <div class="q-item q-item-division relative-position"><img src="http://www.zaragoza.es/cont/paginas/actividades/imagen/web_320x480px.png" class="q-item-side q-item-side-left q-item-section q-item-image">
                    <div class="q-item-main q-item-section">
                        <div class="q-item-label">Anarchy in the Jota S</div>
                        <div class="q-item-sublabel">SCBCN</div>
                    </div>
                </div>
            </div>
        </div>
        <div>
            <div class="q-card">
                <div class="q-item q-item-division relative-position"><img src="http://www.wtf.es/wtf.png" class="q-item-side q-item-side-left q-item-section q-item-image">
                    <div class="q-item-main q-item-section">
                        <div class="q-item-label">Jotas de Codely TV</div>
                        <div class="q-item-sublabel">Around the world</div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
`;
